src/Application/Actions/Hooks/StripePaymentsUpdate.php

```diff
@@ -181,6 +187,63 @@ private function handleChargeSucceeded(Event $event, Response $response): Respon
         return $this->respondWithData($response, $charge);
     }
 
+    private function handleChargeUpdated(Event $event, Response $response): Response
+    {
+        /**
+         * @psalm-suppress UndefinedMagicPropertyFetch
+         * @var Charge $charge
+         */
+        $charge = $event->data->object;
+        if ($charge->status !== Charge::STATUS_SUCCEEDED) {
+            // we're only interested in updates for successful charges.
+            return $this->respond($response, new ActionPayload(204));
+        }
+
+        $intentId = $charge->payment_intent;
+
+        $this->entityManager->beginTransaction();
+
+        /** @var Donation $donation */
+        if (is_string($intentId)) {
+            $donation = $this->donationRepository->findAndLockOneBy(['transactionId' => $intentId]);
+        } else {
+            $donation = null;
+        }
+
+        if (!$donation) {
+            $this->logger->notice(sprintf('Donation not found with Payment Intent ID %s', $intentId ?? 'null'));
+            $this->entityManager->rollback();
+
+            return $this->respond($response, new ActionPayload(204));
+        }
+
+        $donationService = $this->donationService;
+        $donationService->updateDonationStatusFromSuccessfulCharge($charge, $donation);
+
+        $payment_method = $charge->payment_method;
+
+        // given that this is a successful charge it must have been paid by some method.
+        assert($payment_method !== null);
+
+        $this->regularGivingService->updatePossibleMandateFromSuccessfulCharge(
+            $donation,
+            StripePaymentMethodId::of($payment_method)
+        );
+
+        $this->logger->info(sprintf(
+            'Set donation %s Collected based on hook for charge ID %s',
+            $donation->getUuid(),
+            $charge->id,
+        ));
+
+        $this->entityManager->persist($donation);
+        $this->entityManager->commit();
+        $this->entityManager->flush();
+        $this->queueSalesforceUpdate($donation);
+
+        return $this->respondWithData($response, $charge);
+    }
+
```

Comment:
There's quite a bit of duplicated code in this file (and was already). I'd like to tidy that up but if I do it in this PR now I think it will create an awkward merge conflict with the other open matchbot PR.